### ISSUE TITLE
WARNING:tensorflow:Gradients do not exist for variables ['crf/transitions:0'] when minimizing the loss.

### ISSUE THREAD

**Issue by in-the-wind** [xuxingya/tf2crf#15]
hello, sir
I'm trying to directly use the crf.py, and I pass my inputs without anything others.
![image](https://user-images.githubusercontent.com/48121278/96449802-f6f28900-1247-11eb-80b3-df3d3d9c38ff.png)
![image](https://user-images.githubusercontent.com/48121278/96449887-1093d080-1248-11eb-9f05-0869e34da802.png)

When I start training my model, there is a warning saying that "WARNING:tensorflow:Gradients do not exist for variables ['joint_intent_slot/crf/transitions:0'] when minimizing the loss."

I wanna know whether it's abnormal to show this.

**Comment by xuxingya** [comment]
I implemented a test:
```
from tensorflow.keras.layers import Input, Embedding, Bidirectional, GRU, Dense, Dropout
from tensorflow.keras.models import Model
from transformers import TFBertModel
import tensorflow as tf
from tf2crf import CRF

tf.random.set_seed(200)


class TestModel(tf.keras.Model):
    def __init__(self):
        super().__init__(name='intent')
        self.bert = TFBertModel.from_pretrained('bert-base-uncased')
        self.dropout = Dropout(0.1)
        self.dense = Dense(9, activation='relu')
        self.crf = CRF(9)

    def call(self, inputs, **kwargs):
        sequence_out, pooled_output = self.bert(inputs, **kwargs)
        sequence_out = self.dropout(sequence_out)
        sequence_out = self.dense(sequence_out)
        output = self.crf(sequence_out)
        return output


def test():
    model = TestModel()
    model.compile(loss=model.crf.loss, optimizer='adam', metrics=[model.crf.accuracy])

    x = [[5, 2, 3] * 3] * 10
    y = [[1, 2, 3] * 3] * 10

    model.fit(x=x, y=y, epochs=10, batch_size=4)
    model.save('model')


if __name__ == '__main__':
    test()
```
I run the training on my CPU and get these warning:
'WARNING:tensorflow:Gradients do not exist for variables ['tf_bert_model/bert/pooler/dense/kernel:0', 'tf_bert_model/bert/pooler/dense/bias:0'] when minimizing the loss.'
I also found similar warning when training on GPU.
I guess this is a problem of tensorflow.
If you can train normarlly, I think you can just ignore it.

**Comment by in-the-wind** [comment]
> I implemented a test:
> 
> ```
> from tensorflow.keras.layers import Input, Embedding, Bidirectional, GRU, Dense, Dropout
> from tensorflow.keras.models import Model
> from transformers import TFBertModel
> import tensorflow as tf
> from tf2crf import CRF
> 
> tf.random.set_seed(200)
> 
> 
> class TestModel(tf.keras.Model):
>     def __init__(self):
>         super().__init__(name='intent')
>         self.bert = TFBertModel.from_pretrained('bert-base-uncased')
>         self.dropout = Dropout(0.1)
>         self.dense = Dense(9, activation='relu')
>         self.crf = CRF(9)
> 
>     def call(self, inputs, **kwargs):
>         sequence_out, pooled_output = self.bert(inputs, **kwargs)
>         sequence_out = self.dropout(sequence_out)
>         sequence_out = self.dense(sequence_out)
>         output = self.crf(sequence_out)
>         return output
> 
> 
> def test():
>     model = TestModel()
>     model.compile(loss=model.crf.loss, optimizer='adam', metrics=[model.crf.accuracy])
> 
>     x = [[5, 2, 3] * 3] * 10
>     y = [[1, 2, 3] * 3] * 10
> 
>     model.fit(x=x, y=y, epochs=10, batch_size=4)
>     model.save('model')
> 
> 
> if __name__ == '__main__':
>     test()
> ```
> 
> I run the training on my CPU and get these warning:
> 'WARNING:tensorflow:Gradients do not exist for variables ['tf_bert_model/bert/pooler/dense/kernel:0', 'tf_bert_model/bert/pooler/dense/bias:0'] when minimizing the loss.'
> I also found a similar warning when training on GPU.
> I guess this is a problem of tensorflow.
> If you can train normally, I think you can just ignore it.


well，thanks for your help.
I trained my model normally and find it behaved worse after adding the crf layer compared with without the crf layer.
I just feel a little strange. of course, it may be due to the model itself.

I will close this issue.

**Comment by WzMing325** [comment]
Try to add a regularizer term to the variable which has the 'WARNING'

for example
WARNING:tensorflow:Gradients do not exist for variables ['tf_bert_model/bert/pooler/dense/kernel:0
add a regularizer to this kernel:  tf_bert_model/bert/pooler/dense/kernel:0
